contributing/building-a-bento-amp-extension.md

```diff
@@ -0,0 +1,681 @@
+  AMP.registerElement('amp-my-element', AmpMyElement, CSS);
+});
+```
+
+Alternatively, you may have defined styles via JSS that are compiled into CSS and attached via the shadow stylesheet. To tell AMP which class to use, override the `shadowCss` property:
+
+```javascript
+import {CSS} from './my-element.jss';
+
+/** @override */
+AmpMyElement['shadowCss'] = CSS;
+```
+
+## Actions and events
+
+AMP provides a framework for [elements to fire their own
+events](https://github.com/ampproject/amphtml/blob/master/spec/amp-actions-and-events.md)
+to allow users of that element to listen and react to the events. For
+example, the `amp-base-carousel` extension fires a `slideChange` event.
+This allow publishers to listen to that event and react to it, for example, by updating an `amp-selector` state to match the current slide being shown.
+
+The other part of the event-system in AMP is actions. When listening to
+an event on an element usually you'd like to trigger an action (possibly
+on other elements). For example, in the example above, the publisher is
+executing the `toggle` action on `amp-selector`.
+
+The syntax for using this on elements is as follow:
+
+```html
+<amp-base-carousel
+  on="slideChange:my-selector.toggle(index=e.index, value=true)"
+></amp-base-carousel>
+```
+
+To fire events on your element use AMP's action service and the
+`.trigger` method.
+
+```javascript
+actionServiceForDoc(doc.documentElement).trigger(
+  this.element_,
+  'slideChange',
+  createCustomEvent(
+    win,
+    `amp-base-carousel.${name}`,
+    dict({'index': index})
+  ),
+  ActionTrust.DEFAULT
+);
+```
+
+And to expose actions use `registerApiAction` method that your element
+inherits from `PreactBaseElement`. Note, this should correspond directly with the API exposed in the corresponding Preact compnent via `useImperativeHandle`, and the component should be defined using a `forwardRef` accordingly.
+
+```javascript
+this.registerApiAction('close', this.close.bind(this));
+```
+
+Your element could also choose to override the `activate` method
+inherited from BaseElement that would define the default action for your
+element. For example `amp-lightbox` overrides `activate` to define the `open`
+default case.
+
+Make sure your element documentation documents the events and actions it
+exposes both in its own documentation and in [`AMP Actions and Events`](https://github.com/ampproject/amphtml/blob/master/spec/amp-actions-and-events.md).
+
+## Allowing proper validations
+
+One of AMP's features is that a document can be checked against
+validation rules to confirm it's AMP-valid. While your Bento component can be used in non-AMP contexts, it should also smoothly integrate with AMP documents. When you implement your element, the AMP validator needs to be updated to add rules for your element to keep documents using your element valid. In order to do that you need
+to file an issue on the GitHub repo select "Related to: Validator" and
```

Comment:
AFAICT this is not a valid label anymore

Reply:
@ampproject/wg-caching Would you like to update this guidance? It looks like this doc encourages filing an issue to get your working group to write necessary validation rules. Would you like to keep this model, move to a model where folks are encouraged to contribute these on their own and you are added as reviewers, or <insert other options>? If the first option, what is the appropriate issue label?

Reply:
@rcebulko is right, that isn't a valid label anymore. We use "wg: caching." I'm also unsure if most contributors can add labels so an @ mention would be better.

@caroqliu, we created a document on contributing validator rules for [new components](https://github.com/ampproject/amphtml/blob/master/contributing/component-validator-rules.md). Perhaps we can use that here if the content of that document makes sense for Bento. If you're looking for language used previously check out [allowing proper validations](https://github.com/ampproject/amphtml/blob/master/contributing/building-an-amp-extension.md#allowing-proper-validations).

Reply:
Updated here, thank you!